src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1515 @@
+                     int mode)
+{
+  int retval;
+  ARKodeLSRKStepMem step_mem;
+
+  /* access ARKodeLSRKStepMem structure */
+  retval = lsrkStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* perform RHS functions contingent on 'mode' argument */
+  switch (mode)
+  {
+  case ARK_FULLRHS_START:
+
+    /* compute the RHS */
+    if (!(ark_mem->fn_is_current))
+    {
+      retval = step_mem->fe(t, y, step_mem->Fe[0], ark_mem->user_data);
+      step_mem->nfe++;
+      if (retval != 0)
+      {
+        arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
+                        MSG_ARK_RHSFUNC_FAILED, t);
+        return (ARK_RHSFUNC_FAIL);
+      }
+    }
+
+    /* copy RHS vector into output */
+    N_VScale(ONE, step_mem->Fe[0], f);
+
+    break;
+
+  case ARK_FULLRHS_END:
+    /* No further action is needed since the currently
+    available methods evaluate the RHS at the end of each time step*/
+
+    break;
+
+  case ARK_FULLRHS_OTHER:
+
+    /* call f */
+    retval = step_mem->fe(t, y, f, ark_mem->user_data);
+    step_mem->nfe++;
+    if (retval != 0)
+    {
+      arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
+                      MSG_ARK_RHSFUNC_FAILED, t);
+      return (ARK_RHSFUNC_FAIL);
+    }
+
+    break;
+
+  default:
+    /* return with RHS failure if unknown mode is passed */
+    arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
+                    "Unknown full RHS mode");
+    return (ARK_RHSFUNC_FAIL);
+  }
+
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_TakeStepRKC:
+
+  This routine serves the primary purpose of the LSRKStepRKC module:
+  it performs a single RKC step (with embedding, if possible).
+
+  The output variable dsmPtr should contain estimate of the
+  weighted local error if an embedding is present; otherwise it
+  should be 0.
+
+  The input/output variable nflagPtr is used to gauge convergence
+  of any algebraic solvers within the step.  As this routine
+  involves no algebraic solve, it is set to 0 (success).
+
+  The return value from this routine is:
+            0 => step completed successfully
+           >0 => step encountered recoverable failure;
+                 reduce step and retry (if possible)
+           <0 => step encountered unrecoverable failure
+  ---------------------------------------------------------------*/
+int lsrkStep_TakeStepRKC(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+{
+  int retval;
+  sunrealtype* cvals;
+  sunrealtype hmax, w0, w1, temp1, temp2, arg, bjm1, bjm2, mus, thjm1, thjm2, 
+    zjm1, zjm2, dzjm1, dzjm2, d2zjm1, d2zjm2, zj, dzj, d2zj, bj, ajm1, mu, nu, 
+    thj;
+  const sunrealtype onep54 = SUN_RCONST(1.54), c13 = SUN_RCONST(13.0),
+                    p8 = SUN_RCONST(0.8), p4 = SUN_RCONST(0.4);
```

Comment:
A citation on where the come from would be helpful

Reply:
Related citations are on the documentation